File: python/44_WildcardMatching.py
from unittest import TestCase
# https://leetcode.com/problems/wildcard-matching/


class WildcardMatching(TestCase):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        mat = [i == 0 for i in range(len(s)+1)]
        for pi in p:
            pre, mat[0] = mat[0], mat[0] and pi == '*'
            for j, sj in enumerate(s):
                pre, mat[j+1] = mat[j+1], mat[j+1] or mat[j] if pi == '*' else pre and pi in (sj, '?')
        return mat[-1]

    def test1(self):
        self.assertEqual(False, self.isMatch("aa", "a"))

    def test2(self):
        self.assertEqual(True, self.isMatch("aa", "*"))

    def test3(self):
        self.assertEqual(False, self.isMatch("cb", "?a"))

    def test4(self):
        self.assertEqual(True, self.isMatch("abceb", "*a*b"))

    def test5(self):
        self.assertEqual(False, self.isMatch("acdcb", "a*c?b"))

    def test6(self):
        self.assertEqual(False, self.isMatch("aab", "c*a*b"))

File: kotlin/src/main/kotlin/44.WildcardMatching.kt
import org.junit.Assert.assertEquals
import org.junit.Test

//https://leetcode.com/problems/wildcard-matching/

class WildcardMatching{
  fun isMatch(s: String, p: String): Boolean {
    val n = s.length
    val mat = BooleanArray(n+1)
    mat[0] = true

    for(pi in p)
      if(pi == '*') {
        for (j in 0 until n)
          mat[j+1] = mat[j+1] || mat[j]
      }else{
        for(j in n-1 downTo 0)
          mat[j+1] = mat[j] && (pi=='?' || pi==s[j])
        mat[0] = false
      }

    return mat[n]
  }

  @Test
  fun test1(){
    assertEquals(false, isMatch("aa", "a"))
  }

  @Test
  fun test2(){
    assertEquals(true, isMatch("aa", "*"))
  }

  @Test
  fun test3(){
    assertEquals(false, isMatch("cb", "?a"))
  }

  @Test
  fun test4(){
    assertEquals(true, isMatch("adceb", "*a*b"))
  }

  @Test
  fun test5(){
    assertEquals(false, isMatch("acdcb", "a*c?b"))
  }
}